pkg/client/connector/userd_auth/login.go

```diff
@@ -233,12 +233,17 @@ func (l *loginExecutor) Worker(ctx context.Context) error {
 		for {
 			select {
 			case <-l.refreshTimer.C:
-				dlog.Infoln(ctx, "refreshing access token...")
-				if token, err := l.getToken(ctx); err != nil {
-					dlog.Infof(ctx, "could not refresh assess token: %v", err)
-				} else if token != "" {
-					dlog.Infof(ctx, "got new access token")
-				}
+				// Call getToken() in separate goroutine because it may end up
+				// writing to the refreshTimerReset channel that this loop is
+				// the only reader of
```

Comment:
The problem with that is that `l.getToken(ctx)` (may) mutate things in `l` without locking, so doing this in a separate goroutine can corrupt things if a second `<-l.refreshTimer.C` fires while the first goroutine is still running.

I think it would be better to just change the channel to be buffered?

```patch
--- a/pkg/client/connector/userd_auth/login.go
+++ b/pkg/client/connector/userd_auth/login.go
@@ -100,7 +100,10 @@ func NewLoginExecutor(
 		// in a running state with 1 minute left, and then immediately stop it below with
 		// resetRefreshTimerUnlocked.
 		refreshTimer:      time.NewTimer(1 * time.Minute),
-		refreshTimerReset: make(chan time.Duration),
+		// The "refresh" goroutine is the only reader of refreshTimerReset, and
+		// "refresh"->l.getToken()->l.tokenSource.Token()->l.tokenCB() may call
+		// l.resetRefreshTimer() to write to it, so we need a buffer size of at least 1.
+		refreshTimerReset: make(chan time.Duration, 1),
 	}
 	ret.oauth2ConfigMu.Lock()
 	ret.loginMu.Lock()
```